### PR TITLE
Views of entities and components

### LINKED ISSUE #103: Create a View system

### PR DESCRIPTION
**Pull request by AnonymusRaccoon**
## Proposed changes

A new view system that allow optimizations. A view can be fetched like so: 
```cplusplus
wal->scene.view<PositionComponent, CollisionComponent>()
````

Views can be iterated like so:

```cplusplus
for (WAL::Entity &entity : wal->scene.view<PositionComponent>());
for (WAL::ViewEntity<PositionComponent> &entity : wal->scene.view<PositionComponent>());
for (auto &[entity, position] : wal->scene.view<PositionComponent>());
```

Views are resource heavy to create but once they are created, they are cached internally and retrieving them is trivial. Views should not be cached on systems or anywhere as references to views might be invalidaded if a new scene is created.

ViewIterators (things used to iterate views) are invalided if an entity is created, deleted or if a component handled by the view is created or deleted.

Systems now uses views and are templated for this purpose. The onUpdate and onFixedUpdate methods now take a ViewEntity<Dependencies...> for this purpose, a ViewEntity allow you to access the raw Entity with the operator* or the operator-> and has a get<>() method witch allow you to retrieve a component specified in the view. Please use the view get method instead of the entity getComponent. The view allow O(1) access to the component.

## Information
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Non-breaking changes

## Checklist

- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added necessary documentation

Closes #103

### CODE REVIEW

**Review by EternalRat — APPROVED** [review]

**Review by Arthi-chaud — APPROVED** [review]

**Review by GitBluub — APPROVED** [review]

**Review by TrueBabyChaise — APPROVED** [review]

File: lib/wal/sources/Entity/Entity.hpp [hunk 1]
```diff
@@ -25,10 +27,20 @@ namespace WAL
...
 		static unsigned nextID;
+
+		//! @brief Callback called when a component is added
+		//! @param type The type of component
+		void _componentAdded(const std::type_index &type);
+		//! @brief Callback called when a component is removed
+		//! @param type The type of component
+		void _componentRemoved(const std::type_index &type);
+	protected:
+		//! @brief A reference to the ECS.
```
Comment (Octopus773) [inline lib/wal/sources/Entity/Entity.hpp:42]: wth the doc is dans les choux

File: lib/wal/sources/View/View.hpp [hunk 2]
```diff
@@ -0,0 +1,222 @@
...
+		{
+			return std::get<I>(this->_value);
+		}
+	};
+
+	template<typename It, typename ...Components>
+	class ViewIterator
+	{
+	private:
+		It _it;
```
Comment (Octopus773) [inline lib/wal/sources/View/View.hpp:60]: where's the doc

File: lib/wal/sources/View/View.hpp [hunk 3]
```diff
@@ -0,0 +1,222 @@
...
+		{
+			return this->_types;
+		}
+
+		void emplace_back(Entity &entity) override
+		{
+			auto tuple = std::make_tuple<Components *...>(entity.tryGetComponent<Components>()...);
+			if (std::apply([](const auto *...component) {return ((component == nullptr) || ...);}, tuple))
+				return;
+			std::apply([&](auto *...component) {
```
Comment (Octopus773) [inline lib/wal/sources/View/View.hpp:174]: :eyes:

File: sources/Map/Map.cpp [hunk 4]
```diff
@@ -3,12 +3,33 @@
...
 {
+	void MapGenerator::wallCollide(WAL::Entity &entity, const WAL::Entity &wall)
+	{
+		auto *mov = entity.tryGetComponent<MovableComponent>();
+		if (!mov)
+			return;
+		auto &pos = entity.getComponent<PositionComponent>();
+		const auto &wallPos = wall.getComponent<PositionComponent>();
+		auto diff = pos.position + mov->getVelocity() - wallPos.position;
+//		mov->_velocity = Vector3f();
```
Comment (Octopus773) [inline sources/Map/Map.cpp:22]: i'm gonna fix

**Review by Octopus773 — APPROVED** [review]

**Pull request merged**